### DISCUSSION TITLE
Thoughts on salvo

### DISCUSSION THREAD

**Discussion by prabirshrestha** [salvo-rs/salvo#140] [Ideas]
This is similar to https://github.com/trillium-rs/trillium/discussions/117. I have been using trillium but due lack of proper handling have been looking into trying to find a different framework. Before this was using tide so here are some of my thoughts based on migrating my blog to salvo https://github.com/prabirshrestha/rblog/pull/103.

I love trillium apis as it is very simple to understand. Salvo seems to be in similar mindset as trillium.

* I wished salvo didn't have to use `#[handler]` macros. Seems to have been mentioned https://github.com/salvo-rs/salvo/issues/9.
Trillium uses the following api.

```rust
pub async fn hello_world(conn: Conn) -> Conn {
   conn
}
```

This is very similar to the handler function in salvo. 

```rust
#[handler]
pub async fn hello_world(req: &mut Request, depot: &mut Depot, res: &mut Response, ctrl: &mut FlowCtrl) -> Result<()> {
   ...
}
```

Would be good to have similar api as trillium where the `Context` or `Conn` contains request, depot, response and ctrl. This is also similar to a popular go lang framework [echo](https://echo.labstack.com/guide/) and nodejs [koajs](https://github.com/koajs/koa). This makes it easy to transfer knowledge from other frameworks.

```rust
pub async fn hello_world(ctx: &mut Context) -> Result<()> {
}
```

* Wished depot was just called state as most of the frameworks call it. Wasn't mentioned in readme but examples helped here.
* Affix/Depot is a bit confusing coming from trillium. https://docs.trillium.rs/trillium/struct.state
* I had to use affix to regsiter a struct instead of just using depot.

```rust
.hoop(extra::affix::inject(AppState::new_from_env()?))
```

* Wished handlers as tuples worked. Seems it was supported but removed. This makes it easy to read all of the server app. https://github.com/prabirshrestha/rblog/blob/5cb5d742d9cd45c8b3eb27acb28d724a11c03aba/src/app.rs#L9-L33
* Would be good to have more chaining support. 

```rust
res
  .with_headers('key1', 'value1)
  .with_headers('key2', 'value2')
  .with_status(404)
  .with_body("hello");
```

* I like that all I need to do is add salve and preludes and enable extra by features.
* In trillium way to extend everything is by extending `Conn`. If Salvo had `Context` or `Conn` concept this would make it easy. In trillium I have found discovering apis is very easy. 
As mentioned in here I had hard time finding how to send a file. https://github.com/salvo-rs/salvo/issues/137. This could had looked something like this.

```rust
pub async fn hello_world(ctx: &mut Context) -> Result<()> {
  ctx.send_file("./hello.txt")?;
}
```

Another example where this could be useful is when rendering templates. For example in order to port https://github.com/prabirshrestha/trillium-ructe to salvo I don't know what api I should expose. In trillium I just extend the `Conn`, it could had looked like `ctxt.render_html(|o| template::home(o))`.

* Trillium is based on [plug](https://hexdocs.pm/plug/)
* Salvo is based on hyper so the perf is great based on the screenshots shown in readme.

**Comment by chrislearn** [comment]
Thanks for your feedback.

1. I understand some guys do not like macros, But, remove #[handler] macros is hard, can not get many benefits.
2. Different users from different other frameworks and each frameworks has it's own name style, Which one salvo should follow? So salvo will still use current names.
3. Consider to supports handlers as tuple again. https://github.com/salvo-rs/salvo/issues/141
4, Add some chaining support to request and response. https://github.com/salvo-rs/salvo/issues/142
5. Add extra to preludes when features enabled. https://github.com/salvo-rs/salvo/issues/143
6. It is hard to explain why Context or Conn will not be added, the only thing we can do is to improve documentation.

**Reply by prabirshrestha** [reply]
Totally understand on different apis.

As for chaining, should `render` support chaining too?

Thanks for adding the features quickly!

**Reply by chrislearn** [reply]
Done
